stopes/pipelines/asr_bleu/retrieve_data.py

```diff
@@ -0,0 +1,138 @@
+import os
+import logging
+import typing as tp
+import pandas as pd
+from dataclasses import dataclass
+from pathlib import Path
+from glob import glob
+from omegaconf.omegaconf import MISSING
+
+from stopes.core.launcher import Launcher
+from stopes.core.stopes_module import Requirements, StopesModule
+
+@dataclass
+class RetrieveDataJob:
+    audio_path: str = MISSING
+    reference_path: str = MISSING
+    audio_format: str = MISSING
+    reference_format: str = MISSING
+    reference_tsv_column: tp.Optional[str] = None
+
+@dataclass
+class RetrieveDataConfig:
+    retrieve_data_jobs: tp.List[RetrieveDataJob] = MISSING
+
+class RetrieveData(StopesModule):
+    def __init__(self, 
+                 config: RetrieveDataConfig
+    ):
+        print(config)
+        super().__init__(config=config, config_class=RetrieveDataConfig)
+
+    def array(self):
+        return self.config.retrieve_data_jobs
+
+    def requirements(self) -> Requirements:
+        return Requirements(
+            nodes=1,
+            tasks_per_node=1,
+            gpus_per_node=0,
+            cpus_per_task=1,
+            timeout_min=24 * 60,
+        )
+    
+    def _extract_audio_for_eval(self, audio_dirpath: str, audio_format: str):
+        """Extract audio file paths for subsequent transcription"""
+        if audio_format == "n_pred.wav":
+            """
+            The assumption here is that 0_pred.wav corresponds to the reference at line position 0 from the reference manifest
+            """
+            audio_list = []
+            audio_fp_list = glob((Path(audio_dirpath) / "*_pred.wav").as_posix())
+            audio_fp_list = sorted(
+                audio_fp_list, key=lambda x: int(os.path.basename(x).split("_")[0])
+            )
+            for i in range(len(audio_fp_list)):
+                try:
+                    audio_fp = (Path(audio_dirpath) / f"{i}_pred.wav").as_posix()
+                    assert (
+                        audio_fp in audio_fp_list
+                    ), f"{Path(audio_fp).name} does not exist in {audio_dirpath}"
+                except AssertionError:
+                    # check the audio with random speaker
+                    audio_fp = Path(audio_dirpath) / f"{i}_spk*_pred.wav"
+                    audio_fp = glob(
+                        audio_fp.as_posix()
+                    )  # resolve audio filepath with random speaker
+                    assert len(audio_fp) == 1
+                    audio_fp = audio_fp[0]
+
+                audio_list.append(audio_fp)
+        else:
+            raise NotImplementedError
+
+        return audio_list
+    
+    def _extract_text_for_eval(self, references_filepath: str, reference_format: str, reference_tsv_column: str = None):
+        """Extract sentences for reference"""
+        if reference_format == "txt":
+            reference_sentences = open(references_filepath, "r").readlines()
```

Comment:
Please use the `open()` from stopes.core.utils, in general it is good to scour stopes/core/utils.py to see if there are already functions which you can use.